src/storage/src/hummock/iterator/concat_delete_range_iterator.rs

```diff
@@ -0,0 +1,278 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::future::Future;
+
+use risingwave_hummock_sdk::key::{FullKey, PointRange, UserKey};
+use risingwave_hummock_sdk::HummockEpoch;
+use risingwave_pb::hummock::SstableInfo;
+
+use crate::hummock::iterator::DeleteRangeIterator;
+use crate::hummock::sstable_store::SstableStoreRef;
+use crate::hummock::{HummockResult, SstableDeleteRangeIterator};
+use crate::monitor::StoreLocalStatistic;
+
+pub struct ConcatDeleteRangeIterator {
+    sstables: Vec<SstableInfo>,
+    current: Option<SstableDeleteRangeIterator>,
+    idx: usize,
+    sstable_store: SstableStoreRef,
+    stats: StoreLocalStatistic,
+}
+
+impl ConcatDeleteRangeIterator {
+    pub fn new(sstables: Vec<SstableInfo>, sstable_store: SstableStoreRef) -> Self {
+        Self {
+            sstables,
+            sstable_store,
+            stats: StoreLocalStatistic::default(),
+            idx: 0,
+            current: None,
+        }
+    }
+
+    async fn next_inner(&mut self) -> HummockResult<()> {
+        if let Some(iter) = self.current.as_mut() {
+            if iter.is_valid() {
+                if self.idx + 1 < self.sstables.len()
+                    && self.sstables[self.idx + 1].range_tombstone_count > 0
+                    && iter.next_range_epoch() == HummockEpoch::MAX
+                    && iter
+                        .next_extended_user_key()
+                        .left_user_key
+                        .eq(&FullKey::decode(
+                            &self.sstables[self.idx].key_range.as_ref().unwrap().right,
+                        )
+                        .user_key)
+                {
+                    let last_key_in_sst_start =
+                        iter.next_extended_user_key()
+                            .left_user_key
+                            .eq(&FullKey::decode(
+                                &self.sstables[self.idx + 1].key_range.as_ref().unwrap().left,
+                            )
+                            .user_key);
+                    iter.next().await?;
+                    assert!(!iter.is_valid());
+                    if last_key_in_sst_start {
+                        self.seek_idx(self.idx + 1, None).await?;
+                        if self.is_valid()
+                            && self
+                                .next_extended_user_key()
+                                .left_user_key
+                                .eq(&FullKey::decode(
```

Comment:
We should check `self.next_extended_user_key().is_exclude_left_key==false` as well.

Reply:
fixed